### PR TITLE
Add cli_config module

### LINKED ISSUE #39228: [CLI_MOD] cli_config generic module
Platform agnostic module for sending text based config to network devices
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

modules/network/cli/cli_config.py
##### ANSIBLE VERSION

```
2.7
```

### PR DESCRIPTION
**Pull request by trishnaguha**
Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

##### SUMMARY

fixes https://github.com/ansible/ansible/issues/39228

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

modules/network/cli/cli_config.py
##### ANSIBLE VERSION

```
2.7
```

### CODE REVIEW

File: lib/ansible/modules/network/cli/cli_config.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        replace=dict(default=False, type='bool'),
+        rollback=dict(type='int'),
+        commit_comment=dict(type='str'),
+        defaults=dict(default=False, type='bool'),
+        multiline_delimiter=dict(type='str'),
+        diff_replace=dict(choices=['line', 'block', 'config']),
+        diff_match=dict(choices=['line', 'strict', 'exact', 'none']),
+        diff_ignore_lines=dict(type='list'),
+        severity=dict(default='error', choices=['error', 'warning'])
+    )
```
Comment (ganeshrn) [inline lib/ansible/modules/network/cli/cli_config.py]: You can also add option the enable/disable logging of command and it's response in output using [enable_response_logging](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/cliconf/__init__.py#L159) and [disable_response_logging](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/cliconf/__init__.py#L163) api

**Review by Qalthos — APPROVED** [review]

File: lib/ansible/modules/network/cli/cli_config.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+    kwargs = {'candidate': candidate, 'running': running}
+    if diff_match:
+        kwargs.update({'diff_match': diff_match})
+    if diff_replace:
+        kwargs.update({'diff_replace': diff_replace})
+    if diff_ignore_lines:
+        kwargs.update({'diff_ignore_lines': diff_ignore_lines})
+
+    diff_response = connection.get_diff(**kwargs)
```
Comment (ganeshrn) [inline lib/ansible/modules/network/cli/cli_config.py]: get_diff() need to be called only when device doesn't support on-box diff ie. check for value of `supports_onbox_diff` in device capability.

Comment (ganeshrn) [reply lib/ansible/modules/network/cli/cli_config.py]: If device supports onbox diff directly call edit_config() with required parameters the diff part will be taken care by device and if diff is present will be added in the edit_config() resposne

**Review by ganeshrn — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cli/cli_config.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+        if not isinstance(candidate, list):
+            candidate = candidate.strip('\n').splitlines()
+
+        kwargs = {'candidate': candidate, 'commit': commit, 'replace': replace,
+                  'comment': commit_comment}
+        resp = connection.edit_config(**kwargs)
+
+        if 'diff' in resp:
+            if resp['diff']:
+                result['commands'] = resp['diff']
```
Comment (ganeshrn) [inline lib/ansible/modules/network/cli/cli_config.py]: Use module._diff to check if diff is enabled and add the diff to prepared key

**Review by ganeshrn — APPROVED** [review]

**Pull request merged**